crates/oxc_query/src/edges.rs

```diff
@@ -0,0 +1,2170 @@
+use trustfall::provider::{
+    ContextIterator, ContextOutcomeIterator, EdgeParameters, ResolveEdgeInfo, VertexIterator,
+};
+
+use super::vertex::Vertex;
+use crate::Adapter;
+
+pub(super) fn resolve_astnode_edge<'a, 'b: 'a>(
+    contexts: ContextIterator<'a, Vertex<'b>>,
+    edge_name: &str,
+    _parameters: &EdgeParameters,
+    resolve_info: &ResolveEdgeInfo,
+    adapter: &'a Adapter<'b>,
+) -> ContextOutcomeIterator<'a, Vertex<'b>, VertexIterator<'a, Vertex<'b>>> {
+    match edge_name {
+        "ancestor" => astnode::ancestor(contexts, resolve_info, adapter),
+        "parent" => astnode::parent(contexts, resolve_info, adapter),
+        "span" => astnode::span(contexts, resolve_info),
+        _ => {
+            unreachable!("attempted to resolve unexpected edge '{edge_name}' on type 'ASTNode'")
+        }
+    }
+}
+
+fn ancestors<'a, 'b: 'a>(
+    contexts: ContextIterator<'a, Vertex<'b>>,
+    adapter: &'a Adapter<'b>,
+) -> ContextOutcomeIterator<'a, Vertex<'b>, VertexIterator<'a, Vertex<'b>>> {
+    trustfall::provider::resolve_neighbors_with(contexts, |v| {
+        Box::new(
+            Box::new(
+                adapter
+                    .semantic
+                    .nodes()
+                    .ancestors(v.ast_node_id().expect("for vertex to have an ast_node")),
+            )
+            .map(|ancestor| *adapter.semantic.nodes().get_node(ancestor))
+            .map(std::convert::Into::into),
+        )
+    })
+}
+
+fn parents<'a, 'b: 'a>(
+    contexts: ContextIterator<'a, Vertex<'b>>,
+    adapter: &'a Adapter<'b>,
+) -> ContextOutcomeIterator<'a, Vertex<'b>, VertexIterator<'a, Vertex<'b>>> {
+    trustfall::provider::resolve_neighbors_with(contexts, |v| {
+        #[allow(clippy::option_if_let_else)]
+        if let Some(parent) = adapter
+            .semantic
+            .nodes()
+            .parent_node(v.ast_node_id().expect("for vertex to have an ast_node"))
+        {
+            Box::new(std::iter::once((*parent).into()))
```

Comment:
we do this `(*parent).into()` a lot, but it may be confusing, would it be better to do `Vertex::from(*parent)`?

Reply:
It's fine I think